RevokeCgn/handler.ts

```diff
@@ -0,0 +1,142 @@
+import { OrchestratorInput } from "../UpdateCgnOrchestrator";
+import { makeUpdateCgnOrchestratorId } from "../utils/orchestrators";
+import { checkUpdateCgnIsRunning } from "../utils/orchestrators";
+
+type ErrorTypes =
+  | IResponseErrorInternal
+  | IResponseErrorNotFound
+  | IResponseErrorConflict;
+type ReturnTypes =
+  | IResponseSuccessAccepted
+  | IResponseSuccessRedirectToResource<InstanceId, InstanceId>
+  | ErrorTypes;
+
+type IRevokeCgnHandler = (
+  context: Context,
+  fiscalCode: FiscalCode,
+  cgnRevokationRequest: CgnRevokationRequest
+) => Promise<ReturnTypes>;
+
+const checkExistingCgnStatus = (cgnStatus: CgnStatus) =>
+  CgnRevokedStatus.is(cgnStatus) || CgnCanceledStatus.is(cgnStatus)
+    ? fromLeft<IResponseErrorConflict, CgnStatus>(
+        ResponseErrorConflict(
+          "Cannot revoke the user's cgn because it is already revoked or canceled"
+        )
+      )
+    : taskEither.of<IResponseErrorConflict, CgnStatus>(cgnStatus);
+
+export function RevokeCgnHandler(
+  userCgnModel: UserCgnModel
+): IRevokeCgnHandler {
+  return async (context, fiscalCode, revokationReq) => {
+    const client = df.getClient(context);
+    return userCgnModel
+      .findLastVersionByModelId([fiscalCode])
+      .mapLeft<IResponseErrorInternal | IResponseErrorNotFound>(() =>
+        ResponseErrorInternal("Cannot retrieve CGN infos for this user")
+      )
+      .chain(maybeUserCgn =>
+        fromEither(
+          fromOption(
+            ResponseErrorNotFound("Not Found", "User's CGN status not found")
+          )(maybeUserCgn)
+        )
+      )
+      .foldTaskEither<ErrorTypes, CgnStatus>(fromLeft, userCgn =>
+        checkExistingCgnStatus(userCgn.status)
+      )
+      .chain(cgnStatus =>
+        checkUpdateCgnIsRunning(client, fiscalCode, cgnStatus).foldTaskEither<
+          ErrorTypes,
+          | IResponseSuccessAccepted
+          | IResponseSuccessRedirectToResource<InstanceId, InstanceId>
+        >(
+          response =>
+            response.kind === "IResponseSuccessAccepted"
+              ? taskEither.of(response)
+              : fromLeft(response),
+          () =>
+            tryCatch(
+              () =>
+                client.startNew(
+                  "UpdateCgnOrchestrator",
+                  makeUpdateCgnOrchestratorId(fiscalCode, cgnStatus.status),
+                  OrchestratorInput.encode({
+                    fiscalCode,
+                    newStatus: {
+                      motivation: revokationReq.motivation,
+                      revokation_date: new Date(),
+                      status: StatusEnum.REVOKED
+                    }
+                  })
+                ),
+              toError
+            ).bimap(
+              () => ResponseErrorInternal("Cannot call RevokeCgnOrchestrator"),
+              () => {
+                const instanceId: InstanceId = {
+                  id: `${fiscalCode}-${cgnStatus.status}` as NonEmptyString
```

Comment:
isn't this the same as the value returned by  `makeUpdateCgnOrchestratorId` ?

Reply:
It could be the same.